src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java

```diff
@@ -163,58 +158,58 @@
-        this.indexingBufferSize = indexingBufferSize;
-        this.codecName = codecName;
-
-        this.threadPool = threadPool;
+    public InternalEngine(EngineConfig config) throws EngineException {
+        Preconditions.checkNotNull(config.getStore(), "Store must be provided to the engine");
+        Preconditions.checkNotNull(config.getDeletionPolicy(), "Snapshot deletion policy must be provided to the engine");
+        Preconditions.checkNotNull(config.getTranslog(), "Translog must be provided to the engine");
+
+        this.shardId = config.getShardId();
+        this.logger = Loggers.getLogger(getClass(), config.getIndexSettings(), shardId);
+        this.threadPool = config.getThreadPool();
         this.lastDeleteVersionPruneTimeMSec = threadPool.estimatedTimeInMillis();
-        this.indexingService = indexingService;
-        this.warmer = (InternalIndicesWarmer) warmer;
-        this.store = store;
-        this.deletionPolicy = deletionPolicy;
-        this.translog = translog;
-        this.mergePolicyProvider = mergePolicyProvider;
-        this.mergeScheduler = mergeScheduler;
-        this.analysisService = analysisService;
-        this.similarityService = similarityService;
-        this.codecService = codecService;
-        this.compoundOnFlush = compoundOnFlush;
-        this.indexConcurrency = indexConcurrency;
+        this.indexingService = config.getIndexingService();
+        this.warmer = config.getWarmer();
+        this.store = config.getStore();
+        this.deletionPolicy = config.getDeletionPolicy();
+        this.translog = config.getTranslog();
+        this.mergePolicyProvider = config.getMergePolicyProvider();
+        this.mergeScheduler = config.getMergeScheduler();
+        this.analysisService = config.getAnalysisService();
+        this.similarityService = config.getSimilarityService();
+        this.codecService = config.getCodecService();
         this.versionMap = new LiveVersionMap();
-        this.dirtyLocks = new Object[indexConcurrency * 50]; // we multiply it to have enough...
+        this.dirtyLocks = new Object[config.getIndexConcurrency() * 50]; // we multiply it to have enough...
         for (int i = 0; i < dirtyLocks.length; i++) {
             dirtyLocks[i] = new Object();
         }
-        this.optimizeAutoGenerateId = optimizeAutoGenerateId;
-        this.failEngineOnCorruption = failEngineOnCorruption;
-        this.failedEngineListener = failedEngineListener;
+
+        this.mergeSchedulerFailureListener = new FailEngineOnMergeFailure();
+        this.mergeSchedulerListener = new MergeScheduleListener();
+        this.mergeScheduler.addListener(mergeSchedulerListener);
+        this.mergeScheduler.addFailureListener(mergeSchedulerFailureListener);
+
+        this.failedEngineListener = config.getFailedEngineListener();
         throttle = new IndexThrottle();
+        this.config = config;
+        listener = new EngineConfig.ApplySettings(logger, config) {
```

Comment:
afaik the problem only occurs if the ann. inner class is bound in a guice context